README.md

```diff
@@ -1,6 +1,80 @@
+
+## Terminology
+
+1. _Awaitable_ is an object implementing `Interop\Async\Awaitable` and conforming to this specification.
+1. _Value_ is any legal PHP value (including `null`), except an instance of `Interop\Async\Awaitable`.
+1. _Error_ is any value that can be thrown using the `throw` statement.
+1. _Reason_ is an error indicating why an `Awaitable` has failed.
+
+## States
+
+An `Awaitable` MUST be in one of three states: `pending`, `succeeded`, `failed`.
+
+| A promise in … state | &nbsp; |
+|----------------------|--------|
+|`pending`  | <ul><li>MAY transition to either the `succeeded` or `failed` state.</li></ul>                                |
+|`succeeded`| <ul><li>MUST NOT transition to any other state.</li><li>MUST have a value which MUST NOT change.*</li></ul>  |
+|`failed`   | <ul><li>MUST NOT transition to any other state.</li><li>MUST have a reason which MUST NOT change.*</li></ul> |
+
+* _Must not change_ refers to the _reference_ being immutable in case of an object, _not the object itself_ being immutable.
+
+## Consumption
+
+An `Awaitable` MUST implement `Interop\Async\Awaitable` and thus provide a `when` method to access its current or eventual value or reason.
+
+```php
+<?php
+
+namespace Interop\Async;
+
+/**
+ * Representation of a the future value of an asynchronous operation.
+ */
+interface Awaitable
+{
+    /**
+     * Registers a callback to be invoked when the awaitable is resolved.
+     *
+     * @param callable(\Throwable|\Exception|null $exception, mixed $result) $onResolved
+     *
+     * @return void
+     */
+    public function when(callable $onResolved);
+}
+```
+
+The `when` method MUST accept at least one argument:
+
+`$callback` – A callable conforming to the following signature:
+
+```php
+function($error, $value) { /* ... */ }
+```
+
+Any implementation MUST at least provide these two parameters. The implementation MAY extend the `Awaitable` interface with additional parameters passed to the callback. Further arguments to `when` MUST have default values, so `when` can always be called with only one argument. Callbacks MAY decide to consume only one or none of the provided parameters. `when` MUST NOT return a value.
+
+> **NOTE:** The signature doesn't specify a type for `$error`. This is due to the new `Throwable` interface introduced in PHP 7. As this specification is PHP 5 compatible, we can use neither `Throwable` nor `Exception`.
+
+All registered callbacks MUST be executed in the order they were registered. If one of the callbacks throws an `Exception` or `Throwable`, it MUST be rethrown in a callable passed to `Loop::defer` so `Loop::onError` can be properly invoked by the loop. `Loop` refers to the [global event loop accessor](https://github.com/async-interop/event-loop/blob/master/src/Loop.php). The `Awaitable` implementation MUST then continue to call the remaining callbacks with the original reason.
```

Comment:
This is just fine as long as we're not at top-level and using wait() like behavior. Else it will break if we e.g. do a second request on a same object (assuming that object created objects when created).